drivers/ds3231/ds3231.c

```diff
@@ -171,9 +176,32 @@ int ds3231_init(ds3231_t *dev, const ds3231_params_t *params)
         set = CTRL_EOSC;
     }
 
+    /* if configured, enable the interrupts (no SQW output) */
+    if (params->opt & DS3231_OPT_INTER_ENABLE) {
+        set |= CTRL_INTCN;
+    }
+    else {
+        clr |= CTRL_INTCN;
+    }
+
     return _clrset(dev, REG_CTRL, clr, set, 0, 1);
 }
 
+#if IS_USED(MODULE_DS3231_INT)
+int ds3231_init_int(ds3231_t *dev, ds3231_alarm_cb_t cb, void *arg)
```

Comment:
I'm a bit concerned with exposing the raw interrupt callback to the user.
We can't do I2C communication inside interrupt context, but the requirement to do `ds3231_get_alarm_1_flag` / `ds3231_clear_alarm_1_flag` could lead to just that.

How about 

```C
static void _unlock(void *m)
{
    mutex_unlock(m);
}

int ds3231_await_alarm(ds3231_t *dev)
{
    mutex_t mutex = MUTEX_INIT_LOCKED;

    if (gpio_init_int(dev->int_pin, GPIO_IN, GPIO_FALLING, _unlock, &mutex) < 0) {
            return -EIO;
    }

    /* wait for alarm */
    mutex_lock(&mutex);

    uint8_t status, tmp;
    _read(dev, REG_STATUS, &status, 1, 1, 0);

    /* clear interrupt flags */
    tmp = status & ~(STAT_A1F | STAT_A2F);
    res = _write(dev, REG_STATUS, &tmp, 1, 0, 1)
    if (res != 0) {
        return -EIO;
    }

    return status & (STAT_A1F | STAT_A2F);
}
```

Reply:
This is interesting indeed. This literally happened to me a few days ago: tried to clear the flag inside the interrupt => got stuck waiting for a mutex that never came => wasted a few minutes trying to figure out what was going on. Therefore I can understand that it could be considered dangerous. 

Your proposition is fine by me, the only drawback that I see compared to a "user interrupt" is that you would have to block a thread waiting for the mutex while you wait for the alarm. But I'm guessing this is fine since in most cases, this is going to be the desired behavior anyway.

I won't modify the driver today in case someone disagrees with this.

Reply:
OK so it seems everyone agrees with this implementation. 

Quick remark though @benpicco : do we want  `ds3231_await_alarm`to clear the alarm flag ? Is it its job ? User might want to clear it himself later (clearing the flag changes the logic state of the RTC alarm pin) for whatever reason(s). 

Just asking, it doesn't change anything for me other than how the alarm test is going to be implemented.

Reply:
Is there a use case for not clearing the flag?
If the alarm pin is only used to signal the MCU, why would someone want to keep it high after the signal was delivered?

Reply:
In case the alarm pin is not only attached to the MCU but to some other components/IC: syncing multiple MCUs, powering up some external sensors...

I admit those are some very exotic use cases :smiley:

Reply:
Ok, I've updated the driver and the test method, also fixed some "static-tests" issues.